### PR TITLE
Fix minor shell-scripting issues.

### PR DESCRIPTION
**Pull request by drwilly**
fixes to some of my personal pet-peeve shell-scripting issues.

### CODE REVIEW

**Comment by faho** [comment]
There are two reasons why I haven't done this so far:

- All of the `grep | sed` ones should be replaced with `string` - ideally we wouldn't call `grep` or `sed` anywhere.

- I'm not sure if there aren't weird implementations of these tools in the wild - I know that we've previously had issues with BSD sed lacking some features.

The former is moot since you've done this now, and nobody's had the time to write and test it.

The latter is still important, though - have you confirmed that this'll work on GNU and BSD tools at least? `sort -u` seems safe (even Solaris [has it in the xpg version](https://www.unix.com/man-page/opensolaris/1/sort/)!), but I have no idea which `sed` features are portable.

**Comment by drwilly** [comment]
sed is very powerful (turing-complete in fact). I don't see how the string functions you provide would adequately replace it.
wrt. to the posixness: some of the sed calls are non-posix, however in both cases they were non-posix to begin with. Should that be addressed?
`sort -u` is posix

**Comment by faho** [comment]
>sed is very powerful (turing-complete in fact). I don't see how the string functions you provide would adequately replace it.

Ah, but we're not using all of sed! At most, we're using, like you are doing here, grep-like line-matches and s/match/replacement/g. All of that can be done with `string match` and `string replace`.

>wrt. to the posixness: some of the sed calls are non-posix, however in both cases they were non-posix to begin with.

So, we don't strictly care about everything being POSIX. If it is, that's nice and it can usually go in. But what we really care about is compatibility with existing platforms. When all of them extend POSIX in the same way, we can add non-POSIX stuff. Or when we have different ways to do it on different platforms, especially for optional features (e.g. our clipboard stuff uses pbpaste on macOS and xsel elsewhere).

Some things we've tripped over with `sed` in particular:

- `sed -E` for extended regular expressions (which is in POSIX now, but old Centos didn't support it, so presumably it's a semi-recent addition) - #2305

- `\n` (and IIRC `\t`) wasn't supported in the replacement (but was in the match) on macOS. I don't know if that's explicitly in POSIX - (1fa0c4d4d3ea5de180c3f64aa00c478c90f04904)

>sort -u is posix

Yeah, it can go in.

**Comment by drwilly** [comment]
No, we're not using all of sed. Does `string replace` support capture-groups? Not as far as I can see; and no, 2 calls to `replace` are no adequate substitution. So it really doesn't take much looking to find a frequently used use-case where a pure fish implementation would require quite a bit of boilerplate code.

**Comment by faho** [comment]
>Does string replace support capture-groups?

Sure it does. The regex-mode is based on pcre2, so it's quite full-featured. As a rule of thumb, characters are usually "magic", so capturing groups are just `(expression)`, and non-capturing groups are `(?:expression)`. See the [pcre2 syntax specification](http://pcre.org/current/doc/html/pcre2syntax.html) (we might want to add some of that to string's help page).

As a contrived example:

```fish
string replace -r '(.*)=(.*)' '$1\t$2'
```

There's also the "--filter"/"-f" option to remove all lines/arguments where no substitution occurs, and the "-a"/"--all" option which is like sed's "/g".

>and no, 2 calls to replace are no adequate substitution.

They quite often are! They don't have the overhead of `fork()`ing, so spinning up another is rather cheap.

I guess it's also a bit of a stylistic question, but I often find `sed` (and e.g. `awk`) to be a bit dense, and that by using a couple of statements/lines with `string` et al you can make it more readable.

>So it really doesn't take much looking to find a frequently used use-case where a pure fish implementation would require quite a bit of boilerplate code.

I'm not saying that there never is a use-case for non-builtins, but most of what we do is simple match-and-substitute, and for that `string` is usually perfectly capable. And if it isn't, there's the option of enhancing it.

(Also note that the maven example actually parses xml with regex, which is strictly speaking impossible)

**Comment by drwilly** [comment]
Ah, so it's mostly `string` documentation that is lacking :-P

File: share/completions/lpadmin.fish [hunk 1]
```diff
@@ -23,10 +23,9 @@ complete -c lpadmin -s o -d 'Sets the binary communications program to use when
 complete -c lpadmin -s o -d 'Sets the error policy to be used when the printer backend is unable to send the job to the printer. ' -xa 'printer-error-policy=abort-job printer-error-policy=retry-job printer-error-policy=retry-current-job printer-error-policy=stop-printer'
 complete -c lpadmin -s o -xa printer-is-shared=true -d 'Sets the destination to shared/published or unshared/unpublished'
 complete -c lpadmin -s o -xa printer-is-shared=false -d 'Sets the destination to shared/published or unshared/unpublished'
-complete -c lpadmin -s o -d 'Sets the IPP operation policy associated with the destination' -xa "printer-policy=(cat /etc/cups/cupsd.conf | grep \<Policy | sed 's/<Policy \(.\+\)>/\1/')"
+complete -c lpadmin -s o -d 'Sets the IPP operation policy associated with the destination' -xa "printer-policy=(string replace -r --filter '<Policy (.*)>' '$1' < /etc/cups/cupsd.conf)"
```
Comment (faho) [inline share/completions/lpadmin.fish]: Unfortunately, this needs a check for file existance/readability, or it'll print an error.

```fish
(test -r /etc/cups/cupsd.conf; and string...
```

**Comment by faho** [comment]
>…actually it might also be my reading comprehension.

I can't really comment on that, but I do know that string's documentation _is_ lacking.

PRs welcome!

**Comment by mqudsi** [comment]
LGTM. I've recently been studying sed portability and it seems to be used safely.

My biggest trip-up was always using `sed -r` to basically unlock the usage of modifiers like `+` - I hadn't realized at the time it could be portably used by simply "escaping" it first (an odd usage of the term, since technically you are "un"escaping it by prefixing it with `\`).

**Comment by faho** [comment]
Merged, sorry for the delay!

**Pull request merged**